File: src/bmsx/machine/memory/access_kind.ts
export const enum MemoryAccessKind {
	Word,
	U8,
	U16LE,
	U32LE,
	F32LE,
	F64LE,
}

export const MEMORY_ACCESS_KIND_NAMES = ['mem', 'mem8', 'mem16le', 'mem32le', 'memf32le', 'memf64le'] as const;

export function getMemoryAccessKindForName(name: string): MemoryAccessKind | null {
	switch (name) {
		case 'mem':
			return MemoryAccessKind.Word;
		case 'mem8':
			return MemoryAccessKind.U8;
		case 'mem16le':
			return MemoryAccessKind.U16LE;
		case 'mem32le':
			return MemoryAccessKind.U32LE;
		case 'memf32le':
			return MemoryAccessKind.F32LE;
		case 'memf64le':
			return MemoryAccessKind.F64LE;
		default:
			return null;
	}
}

File: src/bmsx_cpp/machine/memory/access_kind.cpp
#include "machine/memory/access_kind.h"

#include <stdexcept>

namespace bmsx {

const char* memoryAccessKindName(MemoryAccessKind kind) {
	return MEMORY_ACCESS_KIND_NAMES[static_cast<size_t>(kind)].data();
}

MemoryAccessKind memoryAccessKindForName(std::string_view name) {
	if (name == "mem") return MemoryAccessKind::Word;
	if (name == "mem8") return MemoryAccessKind::U8;
	if (name == "mem16le") return MemoryAccessKind::U16LE;
	if (name == "mem32le") return MemoryAccessKind::U32LE;
	if (name == "memf32le") return MemoryAccessKind::F32LE;
	if (name == "memf64le") return MemoryAccessKind::F64LE;
	throw std::runtime_error("[MemoryAccessKind] Unknown memory access kind name.");
}

bool isMemoryAccessKindName(std::string_view name) {
	return name == "mem"
		|| name == "mem8"
		|| name == "mem16le"
		|| name == "mem32le"
		|| name == "memf32le"
		|| name == "memf64le";
}

} // namespace bmsx
